File: src/constants/cardRarity.js
/**
 * Card rarity enum IDs (smallint). Must match `backend/internal/domain/card_rarity.go`.
 * @readonly
 */
export const CardRarity = Object.freeze({
  Basic: 0,
  Token: 1,
  Common: 2,
  Rare: 3,
  SuperRare: 4,
  Majestic: 5,
  Marvel: 6,
  Legendary: 7,
  Fabled: 8,
  Promo: 9,
});

/** Ordered display names for each ID 0..9. */
export const CARD_RARITY_NAMES = Object.freeze([
  "Basic",
  "Token",
  "Common",
  "Rare",
  "Super Rare",
  "Majestic",
  "Marvel",
  "Legendary",
  "Fabled",
  "Promo",
]);

/**
 * @param {number} id
 * @returns {string | undefined}
 */
export function cardRarityName(id) {
  return CARD_RARITY_NAMES[id];
}

/**
 * @param {number} id
 * @returns {boolean}
 */
export function isValidCardRarityId(id) {
  return Number.isInteger(id) && id >= 0 && id < CARD_RARITY_NAMES.length;
}

File: backend/internal/domain/card_rarity.go
package domain

import "strconv"

// CardRarity identifies a card rarity (persisted as smallint on cards).
// Values must stay in sync with frontend `src/constants/cardRarity.js`.
type CardRarity int16

const (
	CardRarityToken     CardRarity = 0
	CardRarityCommon    CardRarity = 1
	CardRarityRare      CardRarity = 2
	CardRaritySuperRare CardRarity = 3
	CardRarityMajestic  CardRarity = 4
	CardRarityLegendary CardRarity = 5
	CardRarityFabled    CardRarity = 6
	CardRarityPromo     CardRarity = 7
)

var cardRarityNames = map[CardRarity]string{
	CardRarityToken:     "Token",
	CardRarityCommon:    "Common",
	CardRarityRare:      "Rare",
	CardRaritySuperRare: "Super Rare",
	CardRarityMajestic:  "Majestic",
	CardRarityLegendary: "Legendary",
	CardRarityFabled:    "Fabled",
	CardRarityPromo:     "Promo",
}

// Valid reports whether r is a defined CardRarity constant.
func (r CardRarity) Valid() bool {
	_, ok := cardRarityNames[r]
	return ok
}

// String returns the wire/display name ("Token", "Super Rare", …) or "CardRarity("+decimal+")" if unknown.
func (r CardRarity) String() string {
	if name, ok := cardRarityNames[r]; ok {
		return name
	}
	return "CardRarity(" + strconv.FormatInt(int64(r), 10) + ")"
}

// CardRarities returns every defined rarity constant in ascending ID order.
func CardRarities() []CardRarity {
	return []CardRarity{
		CardRarityToken,
		CardRarityCommon,
		CardRarityRare,
		CardRaritySuperRare,
		CardRarityMajestic,
		CardRarityLegendary,
		CardRarityFabled,
		CardRarityPromo,
	}
}
